components/tidb_query_executors/src/projection_executor.rs

```diff
@@ -0,0 +1,527 @@
+}
+
+fn get_schema_from_exprs(child_schema: &[FieldType], exprs: &[RpnExpression]) -> Vec<FieldType> {
+    exprs
+        .iter()
+        .map(|expr: &RpnExpression| -> FieldType { expr.ret_field_type(child_schema).clone() })
+        .collect::<Vec<FieldType>>()
+}
+
+impl<Src: BatchExecutor> BatchProjectionExecutor<Src> {
+    #[cfg(test)]
+    pub fn new_for_test(src: Src, exprs: Vec<RpnExpression>) -> Self {
+        let schema = get_schema_from_exprs(src.schema(), &exprs);
+
+        Self {
+            context: EvalContext::default(),
+            src,
+            schema,
+            exprs,
+        }
+    }
+
+    pub fn new(config: Arc<EvalConfig>, src: Src, exprs_def: Vec<Expr>) -> Result<Self> {
+        let mut exprs = Vec::with_capacity(exprs_def.len());
+        let mut ctx = EvalContext::new(config);
+        for def in exprs_def {
+            exprs.push(RpnExpressionBuilder::build_from_expr_tree(
+                def,
+                &mut ctx,
+                src.schema().len(),
+            )?);
+        }
+        let schema = get_schema_from_exprs(src.schema(), &exprs);
+
+        Ok(Self {
+            context: ctx,
+            src,
+            schema,
+            exprs,
+        })
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchProjectionExecutor<Src> {
+    type StorageStats = Src::StorageStats;
+
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        &self.schema
+    }
+
+    #[inline]
+    fn next_batch(&mut self, scan_rows: usize) -> BatchExecuteResult {
+        let mut src_result = self.src.next_batch(scan_rows);
+        let child_schema = self.src.schema();
+        let mut eval_result = Vec::with_capacity(self.schema().len());
+        let mut is_drained = src_result.is_drained;
+        let mut logical_rows = src_result.logical_rows;
+        let mut warnings = src_result.warnings;
+        let logical_len = logical_rows.len();
+        let physical_len = src_result.physical_columns.rows_len();
+
+        if is_drained.as_ref().is_ok() && logical_len != 0 {
+            for expr_idx in 0..self.exprs.len() {
+                match self.exprs[expr_idx].eval(
+                    &mut self.context,
+                    child_schema,
+                    &mut src_result.physical_columns,
+                    &logical_rows,
+                    logical_len,
+                ) {
+                    Err(e) => {
+                        is_drained = is_drained.and(Err(e));
+                        logical_rows.clear();
+                        break;
+                    }
+                    Ok(col) => {
+                        if col.is_scalar() {
+                            eval_result.push(VectorValue::from_scalar(
+                                col.scalar_value().unwrap(),
+                                physical_len,
+                            ));
+                        } else {
+                            eval_result.push(col.vector_value().unwrap().as_ref().clone());
```

Comment:
Why `as_ref().clone()`? Could we directly transfer ownership?

Reply:
it seems not easily to do so. I added a piece of comment to reflect it.